File: src/main/java/com/github/alexandrecarlton/idea/settings/layout/editor/codestyle/java/MatchingRule.java
package com.github.alexandrecarlton.idea.settings.layout.editor.codestyle.java;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import org.immutables.value.Value;

import java.util.List;
import java.util.Optional;

@Value.Immutable
@JsonDeserialize(as = ImmutableMatchingRule.class)
public interface MatchingRule {

  Optional<MatchingRuleType> type();

  List<MatchingRuleModifier> modifier();

  Optional<String> name();

  Optional<MatchingRuleOrder> order();

}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/layout/editor/codestyle/java/arrangement/MatchingRule.kt
package com.github.alexandrecarlton.idea.settings.layout.editor.codestyle.java.arrangement

data class MatchingRule(

    val type: MatchingRuleType? = null,

    // TODO: Check if this should default to null
    val modifier: List<MatchingRuleModifier> = emptyList(),

    val name: String? = null,

    val order: MatchingRuleOrder?
)
